File: tests/fixtures/multi-lang-project/src/python/shapes.py
"""Shape classes for testing architecture graph."""
import math
from abc import ABC, abstractmethod
from typing import List


class Shape(ABC):
    """Base class for all shapes."""
    
    @abstractmethod
    def area(self) -> float:
        """Calculate the area of the shape."""
        pass
    
    @abstractmethod
    def name(self) -> str:
        """Return the name of the shape."""
        pass


class Rectangle(Shape):
    """Rectangle shape implementation."""
    
    def __init__(self, width: float, height: float):
        self.width = width
        self.height = height
    
    def area(self) -> float:
        return self.width * self.height
    
    def name(self) -> str:
        return "Rectangle"
    
    def get_width(self) -> float:
        return self.width
    
    def get_height(self) -> float:
        return self.height


class Circle(Shape):
    """Circle shape implementation."""
    
    def __init__(self, radius: float):
        self.radius = radius
    
    def area(self) -> float:
        return math.pi * self.radius ** 2
    
    def name(self) -> str:
        return "Circle"
    
    def get_radius(self) -> float:
        return self.radius


class Triangle(Shape):
    """Triangle shape implementation."""
    
    def __init__(self, base: float, height: float):
        self.base = base
        self.height = height
    
    def area(self) -> float:
        return 0.5 * self.base * self.height
    
    def name(self) -> str:
        return "Triangle"


def calculate_total_area(shapes: List[Shape]) -> float:
    """Calculate total area of all shapes."""
    return sum(shape.area() for shape in shapes)


def sort_by_area(shapes: List[Shape]) -> List[Shape]:
    """Sort shapes by area in ascending order."""
    return sorted(shapes, key=lambda s: s.area())

File: tests/fixtures/multi-lang-project/src/rust/shapes.rs
use std::f64::consts::PI;

/// Base trait for all shapes
pub trait Shape {
    fn area(&self) -> f64;
    fn name(&self) -> &str;
}

/// Rectangle struct
pub struct Rectangle {
    width: f64,
    height: f64,
}

impl Rectangle {
    pub fn new(width: f64, height: f64) -> Self {
        Rectangle { width, height }
    }
    
    pub fn get_width(&self) -> f64 {
        self.width
    }
    
    pub fn get_height(&self) -> f64 {
        self.height
    }
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }
    
    fn name(&self) -> &str {
        "Rectangle"
    }
}

/// Circle struct
pub struct Circle {
    radius: f64,
}

impl Circle {
    pub fn new(radius: f64) -> Self {
        Circle { radius }
    }
    
    pub fn get_radius(&self) -> f64 {
        self.radius
    }
}

impl Shape for Circle {
    fn area(&self) -> f64 {
        PI * self.radius * self.radius
    }
    
    fn name(&self) -> &str {
        "Circle"
    }
}

/// Triangle struct
pub struct Triangle {
    base: f64,
    height: f64,
}

impl Triangle {
    pub fn new(base: f64, height: f64) -> Self {
        Triangle { base, height }
    }
}

impl Shape for Triangle {
    fn area(&self) -> f64 {
        0.5 * self.base * self.height
    }
    
    fn name(&self) -> &str {
        "Triangle"
    }
}

/// Calculate total area of shapes
pub fn calculate_total_area<T: Shape>(shapes: &[T]) -> f64 {
    shapes.iter().map(|s| s.area()).sum()
}

/// Sort shapes by area
pub fn sort_by_area<T: Shape>(shapes: &mut [T]) {
    shapes.sort_by(|a, b| a.area().partial_cmp(&b.area()).unwrap());
}
